core/drivers/crypto/stm32/ecc.c

```diff
@@ -0,0 +1,506 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2021, STMicroelectronics - All Rights Reserved
+ */
+
+#include <drvcrypt.h>
+#include <drvcrypt_acipher.h>
+#include <initcall.h>
+#include <stdlib_ext.h>
+#include <string.h>
+#include <tee_api_types.h>
+
+#include "common.h"
+#include "stm32_pka.h"
+
+static TEE_Result algo_to_pka_cid(uint32_t algo,
+				  enum stm32_pka_curve_id *cid)
+{
+	switch (algo) {
+	case TEE_ALG_ECDSA_P192:
+		*cid = PKA_NIST_P192;
+		break;
+	case TEE_ALG_ECDSA_P224:
+		*cid = PKA_NIST_P224;
+		break;
+	case TEE_ALG_ECDSA_P256:
+		*cid = PKA_NIST_P256;
+		break;
+	case TEE_ALG_ECDSA_P384:
+		*cid = PKA_NIST_P384;
+		break;
+	case TEE_ALG_ECDSA_P521:
+		*cid = PKA_NIST_P521;
+		break;
+	default:
+		EMSG("algorithm %#"PRIx32" not enabled", algo);
+		return TEE_ERROR_NOT_SUPPORTED;
+	}
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result curve_to_pka_cid(uint32_t curve,
+				   enum stm32_pka_curve_id *cid)
+{
+	switch (curve) {
+	case TEE_ECC_CURVE_NIST_P192:
+		*cid = PKA_NIST_P192;
+		break;
+	case TEE_ECC_CURVE_NIST_P224:
+		*cid = PKA_NIST_P224;
+		break;
+	case TEE_ECC_CURVE_NIST_P256:
+		*cid = PKA_NIST_P256;
+		break;
+	case TEE_ECC_CURVE_NIST_P384:
+		*cid = PKA_NIST_P384;
+		break;
+	case TEE_ECC_CURVE_NIST_P521:
+		*cid = PKA_NIST_P521;
+		break;
+	default:
+		EMSG("curve %#"PRIx32" not enabled", curve);
+		return TEE_ERROR_NOT_SUPPORTED;
+	}
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result stm32_gen_keypair(struct ecc_keypair *key, size_t size_bits)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	struct stm32_pka_bn d = { };
+	struct stm32_pka_point pk = { };
+	enum stm32_pka_curve_id cid = PKA_LAST_CID;
+	size_t bytes = 0;
+
+	if (!key)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	FMSG("Using PKA");
+	res = curve_to_pka_cid(key->curve, &cid);
+	if (res)
+		return res;
+
+	res = stm32_pka_get_max_size(&bytes, NULL, cid);
+	if (res)
+		return res;
+
+	if (size_bits > bytes * 8 ||
```

Comment:
What is the use of size_bits variable? the size seems to be related to bytes?

Reply:
It's the size in bits of the key to allocate given by the framework so when compared to the max size return in byte, I have to multiply by 8